clamav/clamav.go

```diff
@@ -0,0 +1,76 @@
+package clamav
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+
+	"github.com/dutchcoders/go-clamd"
+	"gitlab.com/NebulousLabs/errors"
+)
+
+// ClamAV is a client that allows scanning of content for malware.
+type ClamAV struct {
+	staticClam   *clamd.Clamd
+	staticPortal string
+}
+
+// New creates a new ClamAV client that will try to connect to the ClamAV
+// service listening on a TCP socket at the given address and port. Before
+// returning the client, New verifies the connection to ClamAV.
+func New(clamIP, clamPort, portal string) (*ClamAV, error) {
+	var err error
+	defer func() {
+		if err1 := recover(); err1 != nil {
+			err2 := errors.New(fmt.Sprintf("error while trying to connect to ClamAV: %v", err1))
+			err = errors.Compose(err, err2)
+		}
+	}()
+	clam := &ClamAV{
+		staticClam:   clamd.NewClamd(fmt.Sprintf("tcp://%s:%s", clamIP, clamPort)),
+		staticPortal: portal,
+	}
+	err = clam.Ping()
+	if err != nil {
+		return nil, err
+	}
+	return clam, nil
+}
+
+// Ping checks the ClamAV  daemon's state.
+func (c *ClamAV) Ping() error {
+	return c.staticClam.Ping()
+}
+
+// Scan streams the content of the reader to ClamAV for malware scanning.
+// It returns an `infected` flag, a description of the detected malware and an
+// error.
+func (c *ClamAV) Scan(r io.Reader, abort chan bool) (infected bool, description string, err error) {
+	response, err := c.staticClam.ScanStream(r, abort)
+	if err != nil {
+		return
+	}
+	for s := range response {
+		if s.Status == clamd.RES_FOUND {
+			return true, s.Description, nil
+		}
+	}
+	return
+}
+
+// ScanSkylink downloads the content of the given skylink and streams it to
+// ClamAV for scanning. It returns an `infected` flag, a description of the
+// detected malware and an error.
+func (c *ClamAV) ScanSkylink(skylink string, abort chan bool) (infected bool, description string, err error) {
+	resp, err := http.Get(fmt.Sprintf("%s/%s", c.staticPortal, skylink))
```

Comment:
1. Is there a reason why you are not using our go `client` lib for talking to skynet?
2. should we chunk the scanning process? e.g. if we do it 100MB at a time, we can update a timestamp after every chunk and make sure we don't hit the 1 hour timeout in case we are scanning a huge file. We can also pick up a large scan from the middle instead of restarting it.

Reply:
1. We don't necessarily want to tie the malware scanning to a running `skyd` on the same server. There is no need for us to introduce such a restriction.
2. The idea is to not scan files that big. I believe the largest file we want to scan right now is 100MB.

Reply:
From Discord:
* We'll scan the first 100MB for the moment but we'll mark the records that we don't fully scan.
* We'll also mark records where we don't try all possible offsets, so we can re-scan when needed.